stats/legend.cpp

```diff
@@ -0,0 +1,125 @@
+// SPDX-License-Identifier: GPL-2.0
+#include "legend.h"
+#include "statscolors.h"
+#include <QFontMetrics>
+#include <QGraphicsSceneMouseEvent>
+#include <QGraphicsWidget>
+#include <QPen>
+
+static const double legendBorderSize = 2.0;
+static const double legendBoxBorderSize = 1.0;
+static const double legendBoxScale = 0.8;		// 1.0: text-height of the used font
+static const double legendInternalBorderSize = 2.0;
+static const QColor legendColor(0x00, 0x8e, 0xcc, 192); // Note: fourth argument is opacity
+static const QColor legendBorderColor(Qt::black);
+
+Legend::Legend(QGraphicsWidget *chart, const std::vector<QString> &names) :
+	QGraphicsRectItem(chart), chart(chart), displayedItems(0), width(0.0), height(0.0)
+{
+	setZValue(30.0); // On top of everything else.
+	entries.reserve(names.size());
+	int idx = 0;
+	for (const QString &name: names)
+		entries.emplace_back(name, idx++, (int)names.size(), this);
+
+	// Calculate the height and width of the elements
+	if (!entries.empty()) {
+		QFontMetrics fm(entries[0].text->font());
+		fontHeight = fm.height();
+		for (Entry &e: entries)
+			e.width = fontHeight + 2.0 * legendBoxBorderSize +
+				  fm.size(Qt::TextSingleLine, e.text->text()).width();
+	} else {
+		fontHeight = 0.0;
+	}
+	setPen(QPen(legendBorderColor, legendBorderSize));
+	setBrush(QBrush(legendColor));
+
+	resize(); // Draw initial legend
+}
+
+Legend::Entry::Entry(const QString &name, int idx, int numBins, QGraphicsItem *parent) :
+	rect(new QGraphicsRectItem(parent)),
+	text(new QGraphicsSimpleTextItem(name, parent))
+{
+	rect->setZValue(30.0);
+	rect->setPen(QPen(legendBorderColor, legendBoxBorderSize));
+	rect->setBrush(QBrush(binColor(idx, numBins)));
+	text->setZValue(30.0);
+	text->setBrush(QBrush(darkLabelColor));
```

Comment:
I realized later that this was a bad choice in dark mode. But we can fix that once everything is merged, I don't want to hold this up here.